### ISSUE TITLE
Sporadically high "SSL Negotiation" times in new Linux agents

### ISSUE THREAD

**Issue by wildlyinaccurate** [WPO-Foundation/webpagetest#1106]
I've noticed that "SSL Negotiation" times are sometimes very high (~3x higher than usual) in Chrome on the EC2 agents. I can't replicate it on the other agents ~or in Firefox~ (edit: I can replicate it in Firefox now).

Some example test runs:

 * [Normal negotiation times](https://www.webpagetest.org/result/180221_5V_5da4e1fe74321ca5b5023a9713f8bd18/3/details/#waterfall_view_step1) and [another one](https://www.webpagetest.org/result/180221_5V_5da4e1fe74321ca5b5023a9713f8bd18/1/details/#waterfall_view_step1) for good measure
 * [High SSL negotiation time](https://www.webpagetest.org/result/180221_5V_5da4e1fe74321ca5b5023a9713f8bd18/2/details/#waterfall_view_step1)

Let me know if there's any more information I can provide or testing that I can do.

**Comment by wildlyinaccurate** [comment]
Initially I thought this was something to do with switching from the old wptdriver to the new wptagent, but now I'm not so sure. In the graph below, the switch to the new agents happened on 13 December 2017, but the sporadic SSL negotiation times didn't start happening until 16 January 2018.

![screenshot from 2018-02-22 12-05-50](https://user-images.githubusercontent.com/794263/36510683-c4de1988-17c8-11e8-987a-cb33ccbfebe0.png)

I'm also performing some other tests and finding that I can't reproduce the long negotiation times on every site. Some sites have consistently normal times.

**Comment by pmeenan** [comment]
Looks like Chrome 64 didn't go out until January 24th (in case it had to do with Chrome)

**Comment by pmeenan** [comment]
Might be worth enabling tcpdump to see if we can capture what is going on during the slower negotiations in case there is something not showing up in the waterfall.

**Comment by wildlyinaccurate** [comment]
Sure! Here's a tcpdump of the [normal times](https://www.dropbox.com/s/sgmfja99s5zk5s4/normal-ssl.cap?dl=0) and here's one of the [slow times](https://www.dropbox.com/s/lwtyiyjy6bfr4gj/slow-ssl.cap?dl=0).

Here's a screencap of the certificate exchange in the slow one. A retransmission so early in the exchange is a little bit suspect. I wonder if this is an EC2 networking issue - do you have any non-EC2 Linux agents running on WPT that I could run tests on?

![screenshot from 2018-02-22 12-43-18](https://user-images.githubusercontent.com/794263/36512310-c47e8008-17ce-11e8-95d4-c2bb56220e27.png)

**Comment by wildlyinaccurate** [comment]
I was able to reproduce it on Dulles, although it seems to happen _much_ less often. Here's a test where you can see it happening: https://www.webpagetest.org/result/180226_K0_3578e48ac0493c01148fd1b22d255545/4/details/#waterfall_view_step1

**Comment by wildlyinaccurate** [comment]
I can also reproduce it on Firefox. At this point I'm not really sure how to proceed with the investigation. I'm definitely no expert on this stuff so I could be completely missing the mark but here's a brain dump:

 - The client hangs during the TLS handshake, right after the server key exchange.
 - This causes the server to retransmit a packet after what looks like a 300ms delay. In most cases the client sends the packet not long after the 300ms delay.
 - The CPU is not idle during the delay, which makes me wonder whether the client is taking a long time to verify the server key?
 - I don't think misconfigured firewall rules are to blame, because the handshake completes eventually.
 - I wonder whether the client's crypto lib has been configured to perform the verification in a blocking way?
 - I don't see any evidence of this in the tcpdump but could the client be blocking the handshake while it fetches an up-to-date certificate revocation list?

**Comment by pmeenan** [comment]
I don't think it is a OCSP/CRL check - the tcpdump doesn't show it.

I managed to reproduce it with Chrome's netlog enabled (run 6 [here](https://www.webpagetest.org/result/180226_42_52e7df1453c3d3d5631bd70180e5e69f/)). With a 150ms RTT it had a 577ms SSL negotiation and the netlog shows 578ms so 

I stripped out the raw hex bytes of the payload but the key events look to be:
```
t= 3165 [st=    0] +SOCKET_ALIVE  [dt=10766]
t= 3165 [st=    0]   +TCP_CONNECT  [dt=152]
t= 3165 [st=    0]      TCP_CONNECT_ATTEMPT  [dt=152]
                        --> address = "104.28.12.62:443"
t= 3317 [st=  152]   -TCP_CONNECT
t= 3317 [st=  152]   +SOCKET_IN_USE  [dt=10614]
t= 3317 [st=  152]     +SSL_CONNECT  [dt=578]
t= 3318 [st=  153]        SSL_HANDSHAKE_MESSAGE_SENT
t= 3318 [st=  153]        SOCKET_BYTES_SENT
                          --> byte_count = 208
t= 3475 [st=  310]        SOCKET_BYTES_RECEIVED
                          --> byte_count = 1460
t= 3475 [st=  310]        SSL_HANDSHAKE_MESSAGE_RECEIVED
                          --> type = 2
t= 3863 [st=  698]        SOCKET_BYTES_RECEIVED
                          --> byte_count = 3470
t= 3863 [st=  698]        SSL_HANDSHAKE_MESSAGE_RECEIVED
                          --> type = 11
t= 3864 [st=  699]        SSL_HANDSHAKE_MESSAGE_RECEIVED
                          --> type = 22
t= 3864 [st=  699]        SSL_HANDSHAKE_MESSAGE_RECEIVED
                          --> type = 12
t= 3865 [st=  700]        SSL_HANDSHAKE_MESSAGE_RECEIVED
                          --> type = 14
t= 3865 [st=  700]        SSL_HANDSHAKE_MESSAGE_SENT
                          --> type = 16
t= 3865 [st=  700]        SSL_HANDSHAKE_MESSAGE_SENT
                          --> type = 20
t= 3865 [st=  700]        SOCKET_BYTES_SENT
                          --> byte_count = 93
t= 3865 [st=  700]        SSL_CERTIFICATES_RECEIVED
t= 3866 [st=  701]       +CERT_VERIFIER_REQUEST  [dt=27]
t= 3893 [st=  728]       -CERT_VERIFIER_REQUEST
t= 3893 [st=  728]        SIGNED_CERTIFICATE_TIMESTAMPS_RECEIVED
t= 3894 [st=  729]        SIGNED_CERTIFICATE_TIMESTAMPS_CHECKED
t= 3895 [st=  730]        CERT_CT_COMPLIANCE_CHECKED
                          --> build_timely = true
                          --> ct_compliance_status = "COMPLIES_VIA_SCTS"
t= 3895 [st=  730]     -SSL_CONNECT
                        --> cipher_suite = 49195
                        --> is_resumed = false
                        --> next_proto = "h2"
                        --> version = "TLS 1.2"
t= 3895 [st=  730]     +SOCKET_IN_USE  [dt=10036]
```

From the in-browser net stack's perspective, it looks like it got the first part of the TLS handhake from the server ~150ms after sending the client hello but there was ~380ms gap between the first packet's worth of data (1460 byte read) and the remaining 3470 bytes of the server hello.

The tcpdump captures packets outside of the traffic-shaping so there is ~150ms between what you see in the tcpdump and the actual kernel.

The tcpdump doesn't look like it is showing the same delay but it is showing a retransmit:
![image](https://user-images.githubusercontent.com/444165/36735830-1b58b0e0-1ba5-11e8-9d42-29378b8f574a.png)

The ACKs for everything after the first packet get stalled and don't seem to go out until after the retransmit though the retransmit is only for the last bit which is the certificate.  There isn't any other traffic in the tcpdump so it doesn't look like a contention issue.

From the netlog and tcpdump it looks like some interaction between the traffic-shaping and tcp stacks on either end of the connection.  it's still not 100% clear what happened in the 300ms between the ACK for 1461 and the ACK for 2921.  It could be a delayed ACK where ACK's are only sent for every other packet (with up to a 500ms delay) but there also shouldn't have been more than a few milliseconds between 1461 and 2921 arriving (with the 1.6Mbps tc profile).

It's possible that there is some scheduling voodoo going on wither with the VM itself or the accuracy netem is getting with the kernel ticks but neither should be nearly that big (maybe ~4ms of jitter).

I'll try on a few different configs (Linux vs Windows, VM vs real hardware, etc) to see what conditions it appears to show up in.

**Comment by wildlyinaccurate** [comment]
Thanks Pat, really appreciate the detailed response. Let me know if there's any further testing or investigation I can do on my end?

**Comment by pmeenan** [comment]
I have more data but not necessarily answers yet.

* Across 50 runs on the Dulles VM's it showed up [5 times](https://www.webpagetest.org/graph_page_data.php?tests=180228_BB_96f4e53e18631b39cefc8e4a168aed4e&medianMetric=SpeedIndex#basePageSSLTime).
* When I reduced the number of VM's on each server from 10 to 8 it showed up [2 times](https://www.webpagetest.org/graph_page_data.php?tests=180228_X2_2bfe6729e1cb356646e28b15dbc5c64c&medianMetric=SpeedIndex#basePageSSLTime)
* On the physical thinkpads running windows there were [no long outliers](https://www.webpagetest.org/graph_page_data.php?tests=180228_WB_a323b8b3c1dc0c8ce1fafdffc79bd442&medianMetric=SpeedIndex#basePageSSLTime) but the timings were more variable.
* On the c4.large EC2 instance it showed up [14 times](https://www.webpagetest.org/graph_page_data.php?tests=180228_CF_c23a6142105f7338c1b8d79fd3c74e9d&medianMetric=SpeedIndex#basePageSSLTime)
* On a single VM running on my desktop dev machine it was [super clean](https://www.webpagetest.org/graph_page_data.php?tests=180228_0D_0b6d3a5fcb4e6bbc2c6d83c217fb949c&medianMetric=SpeedIndex#basePageSSLTime) with a really tight packing and only a single result that varied by 20ms. That VM is running desktop instead of server though (not sure if it uses a different scheduler).

It feels a lot like a VM scheduling issue but it's still hard to be sure.  I'm going to try reducing the VM stacking on the Dulles Linux agents further to see if it can be completely eliminated or how low it needs to go for that to happen.  That doesn't help with the c4.large results though.  I may try c5.large as they switched from Xen to KVM to see if that impacts anything.

**Comment by pmeenan** [comment]
Running EC2 video capture at 10fps to match Dulles (instead of 30fps) showed [2 outliers](https://www.webpagetest.org/graph_page_data.php?tests=180228_SJ_cf07475591fc14a82f9834ed5cf1083e&medianMetric=SpeedIndex#basePageSSLTime).  It looks like at a minimum there may be some contention with video capture at high frame rates.

I also tested against the actual WPT origin instead of the Fastly edge (to match the Dulles tests) and it was a bit cleaner though there were still [7 outliers](https://www.webpagetest.org/graph_page_data.php?tests=180228_75_20b06c367e7a4b558338cab1044df1bf&medianMetric=SpeedIndex#basePageSSLTime)

I have some test VM's on GCE (standard-2) and they had [1 outlier](https://www.webpagetest.org/graph_page_data.php?tests=180228_MG_45ff73ec49113784ad7e4d91ae7fa922&medianMetric=SpeedIndex#basePageSSLTime) though it's hard to be 100% sure since that VM is on the west coast.

The frame rate can be specified in the user data string as fps=10 or wpt_fps=10 (or on a test-by-test bases with fps=10 as a test parameter).  I may also see about releasing new AMIs that default to 10fps.

Checking now to see if there is anything the agent can do to reduce the contention as well (using nice on ffmpeg for example).

**Comment by wildlyinaccurate** [comment]
This is super interesting! I still see outliers when I uncheck the _Capture Video_ checkbox, but maybe that's not doing what I expect it to do. Do you have any ideas about why this would only happen with wptagent? Would it be worthwhile for me to attempt to diff the wptagent and wptdriver AMIs?

**Comment by pmeenan** [comment]
With the latest changes, the EC2 linux wptagent agents are looking pretty good ([one outlier](https://www.webpagetest.org/graph_page_data.php?tests=180228_BW_f2ad27d8f05cfa6725d3860eed3722e1&medianMetric=SpeedIndex#basePageSSLTime)).

I'm testing out the lowlatency kernel locally right now to see if it cleans up the remaining scheduler outliers.

**Comment by pmeenan** [comment]
The lowlatency kernel may have [reduced the variability](https://www.webpagetest.org/graph_page_data.php?tests=180228_ZV_de78da4edf19b415b925de37dc41b60d&medianMetric=SpeedIndex#basePageSSLTime) a bit but didn't change the outlier behavior.  I'll see about trying it out on EC2 to see if it does better there.

**Comment by pmeenan** [comment]
Not sure it is completely solved but I just pushed out new AMI's with the lowlatency kernel and it's certainly a [lot better](https://www.webpagetest.org/graph_page_data.php?tests=180228_8R_7ff5b6374ee719a212c748c1e7420273&medianMetric=SpeedIndex#basePageSSLTime).  I hedge a bit because those are c5.large because US East doesn't have spot capacity for c4.large right now but it looks [pretty good from California](https://www.webpagetest.org/graph_page_data.php?tests=180228_2S_05001ede0d49115fa755a2bd0d734d1e&medianMetric=SpeedIndex#basePageSSLTime) (c4) as well (one outlier).

I'm in the process of rolling out the lowlatency kernel to the Dulles agents right now but I get the feeling this is about as good as it is going to get.

**Comment by wildlyinaccurate** [comment]
Wow, amazing work! We'll roll out the updated agents and see what sort of impact that has. I'll get back to you once we have enough data to confirm the effect.

**Comment by wildlyinaccurate** [comment]
We've only been running tests on the updated agents for about 24 hours but so far we're not seeing a reduction in variability. Here are some tests from [earlier in the week](https://wpt1.speedcurve.com/graph_page_data.php?tests=180223_NE_143090ff1e40220126e8c13ea7fee2ff,180223_V6_4b50145721364af704598d9bc4fd83e8,180223_QG_0a8df11e47ab0429db7265d2561b1b00,180223_B5_c397da64c044c603aae7fb9cef9c2fbf,180223_3N_3f9f63f4dd127421c914b76d2438ebb5,180223_4G_029b18a66466065520156c2657dcb586,180224_K1_c11130b350923be7c4fa5b8b0893a382,180224_RD_15d18f8744f061ca5f6948d5530df073,180224_56_5f3fb3e424375fffab06fe54d3483391&medianMetric=SpeedIndex#basePageSSLTime) and also some from the [last 24 hours](https://wpt1.speedcurve.com/graph_page_data.php?tests=180301_FN_09623176f826231cda02ef65d6a80719,180301_49_ff01c6bd103a096d67d1edfcdde1f8f6,180301_CD_c31c85e36c2d59f729579b787476db9c,180301_GC_1d090adff312777473722c964bf7b85f,180301_AC_5824986e8eae9c712fb26de3fd0e95a0,180301_TN_2a125e55bc0f415abc1cba48c467fbab,180301_MF_63f3642d0e70e6c8ce33a1aa13f9fded,180301_S9_80a5bf93c5159a91f8617f86379b1961,180301_CX_df74be9c43b4597966eddfb0a98b3edd&medianMetric=SpeedIndex#basePageSSLTime). The tests from earlier in the week are all run from EC2 in ap-southeast-2 and sa-east-1. The tests from the last 24 hours are the same but with some from eu-west-2 just because I wondered whether the region would make a difference (fairly sure it doesn't).

**Comment by pmeenan** [comment]
Argh.  That sound you heard was me screaming.  It seems to be really sensitive to timing because I [don't see it](https://www.webpagetest.org/graph_page_data.php?tests=180302_EQ_f2cb344fe7a355eaaebc09baeed211b1&medianMetric=SpeedIndex#basePageSSLTime) when hitting https://agent.webpagetest.org/ (my origin) but hitting wildlyinaccurate causes it to happen [a LOT](https://www.webpagetest.org/graph_page_data.php?tests=180302_G5_b7880bc87a9a16408d8981354b740756&medianMetric=SpeedIndex#basePageSSLTime) even with the c5 instances.

I looked through a couple of the wildlyinaccurate tcpdumps where it did go wide and they all show the same pause and retransmission of the certificate packet with no real indication of why it chose to retransmit from there (the last ack out was for much earlier in the stream).

Digging into it more now to see if I can find anything.  Looking into the ipfw configs for shaping and the tcp options for the kernel to see if I can find anything to help with it.

**Comment by pmeenan** [comment]
In testing it looks like if I only did latency but not bandwidth shaping it eliminated the issue.  The shaping used 2 filters, netem for latency and packet loss followed by tbf for bandwidth.  I'm trying out a change right now that simplifies it and does the bandwidth shaping in netem as well (I think older versions of netem didn't support rate limiting which is why it was split).

It'll take about an hour for the update to roll out to all of the agents and I'll do some testing again to see if it helped.

**Comment by pmeenan** [comment]
So far it is looking good but I thought that before as well:

wildlyinaccurate:
* EC2 c5.large - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_Z1_ad7f3e74b8887d2808bf1d6e6fa422a0&medianMetric=SpeedIndex#basePageSSLTime)
* EC2 c4.large - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_34_c9d06e6c9336a964054934fc91099822&medianMetric=SpeedIndex#basePageSSLTime)
* Dulles VM - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_R2_085e46c926375a012ce3a0f0fc305b09&medianMetric=SpeedIndex#basePageSSLTime)

webpagetest:
* EC2 c5.large - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_4B_1e06d9a778126623b71cb6b5a35b60a3&medianMetric=SpeedIndex#basePageSSLTime)
* EC2 c4.large - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_79_5a3d0fb8fd9205ff4dbb4146bcab15c5&medianMetric=SpeedIndex#basePageSSLTime)
* Dulles VM - [No outliers](https://www.webpagetest.org/graph_page_data.php?tests=180302_8B_5a99050619d8bfeb22beb8410ff1db4a&medianMetric=SpeedIndex#basePageSSLTime)

There was some slight variation but nothing like it was before.

If you are using the public AMI's they should all be updated now as well so just keep an eye on it for the next few days.

**Comment by wildlyinaccurate** [comment]
I'm _really_ sorry that this issue caused you to scream! The good news is that the last few test runs for a bunch of the sites I was monitoring are looking stable. I'll continue to monitor this weekend and close the issue if it's resolved.

**Comment by pmeenan** [comment]
Sorry, the screaming was directed at the tbf implementation in Linux.  I'm very grateful for the report and continued investigation.  Highly accurate network characteristics are the foundation for most of WebPageTest and I'd much rather know about an issue than not.  Glad to hear it is looking cleaner.  Fingers crossed it holds.

**Comment by wildlyinaccurate** [comment]
When looking at individual URLs it looks like the variability is totally gone and the backend times are much more predictable. I've also tried to aggregate the backend times across several hundred URLs but the data is too noisy. I'm pretty sure that the sample of individual URLs I'm looking at is fair, though (whatever that means) so I'm going to say that this issue is resolved.

Thanks a MILLION for chasing this one down. I learned a lot through your detailed debugging.
